src/machine/machine_stm32f103xx.go

```diff
@@ -290,3 +291,463 @@ func (spi SPI) setPins(sck, mosi, miso uint8) {
+
+			// send address
+			err = i2c.sendAddress(uint8(addr), false)
+			if err != nil {
+				return err
+			}
+
+			// clear address here
+			timeout := TIMEOUT
+			for i2c.Bus.SR2&(stm32.I2C_SR2_MSL|stm32.I2C_SR2_BUSY) == 0 {
+				timeout--
+				if timeout == 0 {
+					return errors.New("I2C timeout on read clear address")
+				}
+			}
+
+			// Enable ACK of received data
+			i2c.Bus.CR1 |= stm32.I2C_CR1_ACK
+
+			// wait for btf. we need a longer timeout here than normal.
+			timeout = 1000
+			for (i2c.Bus.SR1 & stm32.I2C_SR1_BTF) == 0 {
+				timeout--
+				if timeout == 0 {
+					println("I2C timeout on read 3 bytes")
+					return errors.New("I2C timeout on read 3 bytes")
+				}
+			}
+
+			// Disable ACK of received data
+			i2c.Bus.CR1 &^= stm32.I2C_CR1_ACK
+
+			// read the first byte
+			r[0] = byte(i2c.Bus.DR)
+
+			timeout = 1000
+			for (i2c.Bus.SR1 & stm32.I2C_SR1_BTF) == 0 {
+				timeout--
+				if timeout == 0 {
+					return errors.New("I2C timeout on read 3 bytes")
+				}
+			}
+
+			// Generate stop condition
+			i2c.Bus.CR1 |= stm32.I2C_CR1_STOP
+
+			// read the last 2 bytes by reading twice.
+			r[1] = byte(i2c.Bus.DR)
+			r[2] = byte(i2c.Bus.DR)
+
+			// wait for stop
+			return i2c.waitForStop()
+
+		default:
+			// more than 3 bytes of data to read
+
+			// send address
+			err = i2c.sendAddress(uint8(addr), false)
+			if err != nil {
+				return err
+			}
+
+			// clear address here
+			timeout := TIMEOUT
+			for i2c.Bus.SR2&(stm32.I2C_SR2_MSL|stm32.I2C_SR2_BUSY) == 0 {
+				timeout--
+				if timeout == 0 {
+					return errors.New("I2C timeout on read clear address")
+				}
+			}
+
+			for i := 0; i < len(r)-3; i++ {
+				// Enable ACK of received data
+				i2c.Bus.CR1 |= stm32.I2C_CR1_ACK
+
+				// wait for btf. we need a longer timeout here than normal.
+				timeout = 1000
+				for (i2c.Bus.SR1 & stm32.I2C_SR1_BTF) == 0 {
+					timeout--
+					if timeout == 0 {
+						println("I2C timeout on read 3 bytes")
+						return errors.New("I2C timeout on read 3 bytes")
+					}
+				}
+
+				// read the next byte
+				r[i] = byte(i2c.Bus.DR)
+			}
+
+			// wait for btf. we need a longer timeout here than normal.
+			timeout = 1000
+			for (i2c.Bus.SR1 & stm32.I2C_SR1_BTF) == 0 {
+				timeout--
+				if timeout == 0 {
+					return errors.New("I2C timeout on read more than 3 bytes")
+				}
+			}
+
+			// Disable ACK of received data
+			i2c.Bus.CR1 &^= stm32.I2C_CR1_ACK
+
+			// get third from last byte
+			r[len(r)-3] = byte(i2c.Bus.DR)
+
+			// Generate stop condition
+			i2c.Bus.CR1 |= stm32.I2C_CR1_STOP
+
+			// get second from last byte
+			r[len(r)-2] = byte(i2c.Bus.DR)
+
+			timeout = TIMEOUT
+			for (i2c.Bus.SR1 & stm32.I2C_SR1_RxNE) == 0 {
+				timeout--
+				if timeout == 0 {
+					return errors.New("I2C timeout on read last byte of more than 3")
+				}
+			}
+
+			// get last byte
+			r[len(r)-1] = byte(i2c.Bus.DR)
+
+			// wait for stop
+			return i2c.waitForStop()
+		}
+	}
+
+	return nil
+}
+
+const TIMEOUT = 500
```

Comment:
This is in all-caps, meaning it is exported. Also, it is a global constant just for I2C without I2C in the name.
Can you rename it so it at least won't be exported?

Reply:
Fixed.